react-basic-practice/src/features/SignUpForm/index.jsx

```diff
@@ -1,86 +1,77 @@
 
-import REGEXP from '../../constants/regexp';
-
 import Button from '../../components/common/Button';
 import Input from '../../components/common/Input';
 import Logo from '../../components/common/Logo';
 import FormGroup from '../../components/FormGroup';
 import './index.css';
 
+import { createUser, getAllUsers } from '../../services/Users';
+import Validate from '../../helpers/validate';
+import { useLoading } from '../../contexts/loading';
+
 const initialErrorMsgs = {
-  inputEmail: '',
-  inputUserName: '',
-  inputPassword: '',
+  email: '',
+  userName: '',
+  password: '',
+  form: '',
 };
 
 const initialInput = {
-  inputEmail: '',
-  inputUserName: '',
-  inputPassword: '',
+  email: '',
+  userName: '',
+  password: '',
 };
 
-
 const SignUpForm = () => {
-  const [isSignUpLoading, setIsSignUpLoading] = useState(false);
   const [errorMessage, setErrorMessage] = useState(initialErrorMsgs);
   const [inputValue, setInputValue] = useState(initialInput);
+  const navigate = useNavigate();
+  const { loading, show, hide } = useLoading();
 
   const handleInputValue = (value) => {
     setInputValue({ ...inputValue, ...value });
   };
 
-  const validation = () => {
-    for (const key in inputValue) {
-      if (inputValue[key] === '') {
-        setErrorMessage((preMsg) => ({
-          ...preMsg,
-          [key]: 'This input is required.',
-        }));
-      } else {
-        setErrorMessage((preMsg) => ({
-          ...preMsg,
-          [key]: '',
-        }));
-      }
-    }
-  };
-
-  const validateInputs = () => {
-    if (inputValue.inputPassword !== '') {
-      REGEXP.REGEXP_PASSWORD.test(inputValue.inputPassword)
-        ? setErrorMessage((preMsg) => ({ ...preMsg, inputPassword: '' }))
-        : setErrorMessage((preMsg) => ({ ...preMsg, inputPassword: 'Invalid password' }));
-    }
-
-    if (inputValue.inputEmail !== '') {
-      REGEXP.REGEXP_MAIL.test(inputValue.inputEmail)
-        ? setErrorMessage((preMsg) => ({ ...preMsg, inputEmail: '' }))
-        : setErrorMessage((preMsg) => ({ ...preMsg, inputEmail: 'Invalid email' }));
-    }
-
-    if (inputValue.inputUserName !== '') {
-      REGEXP.REGEXP_USER_NAME.test(inputValue.inputUserName)
-        ? setErrorMessage((preMsg) => ({ ...preMsg, inputUserName: '' }))
-        : setErrorMessage((preMsg) => ({ ...preMsg, inputUserName: 'Invalid user name' }));
-    }
-  };
-
-  const handleSignUp = (e) => {
+  const handleSignUp = async (e) => {
     try {
       e.preventDefault();
 
-      setIsSignUpLoading(true);
-
-      validation();
-      validateInputs();
-
-      setIsSignUpLoading(false);
+      show();
+
+      // Check validation input
+      const validateInput = Validate(inputValue);
+      setErrorMessage(validateInput.validateError);
+
+      // Check email already exists.
+      const dataUser = await getAllUsers();
+      const validate = dataUser.some((user) => user.email === inputValue.email);
+
+      if (!validateInput.error) {
```

Comment:
Check input error before checking user exists.

Reply:
The condition is here to make sure that the input value is valid, then send an API call to check whether it is valid or not.

Reply:
If the input value is not valid, then you should return and show error right away. If you can make API calls then the input value is already valid.